numpy/core/src/multiarray/methods.c

```diff
@@ -1008,6 +1008,44 @@ array_getarray(PyArrayObject *self, PyObject *args)
 
 
 static PyObject *
+array_ufunc(PyArrayObject *self, PyObject *args, PyObject *kwds)
+{
+    PyObject *ufunc, *method_name, *normal_args, *ufunc_method, *result;
+
+    if (PyTuple_Size(args) < 2) {
+        PyErr_SetString(PyExc_TypeError,
+                        "__array_ufunc__ requires at least 2 arguments");
+        return NULL;
+    }
+    ufunc = PyTuple_GET_ITEM(args, 0);  /* borrowed reference */
+    if (ufunc == NULL) {
+        return NULL;
+    }
+
+    method_name = PyTuple_GET_ITEM(args, 1);
+    if (method_name == NULL) {
+        return NULL;
+    }
+
+    normal_args = PyTuple_GetSlice(args, 2, PyTuple_GET_SIZE(args));
+    if (normal_args == NULL) {
+        return NULL;
+    }
+
+    ufunc_method = PyObject_GetAttr(ufunc, method_name);
+    if (ufunc_method == NULL) {
+        Py_DECREF(normal_args);
+        return NULL;
+    }
+
+    result = PyObject_Call(ufunc_method, normal_args, kwds);
```

Comment:
The idea here is to add `__array_ufunc__` to ndarray. That base implementation the call the relevant ufunc, which in turn, would normally call `__array_ufunc__` again. To avoid that endless loop, `PyUFunc_CheckOverride` is modified so that it does not make the call when `__array_ufunc__` is the version defined in the ndarray base class.

@mhvk wants the function in the base class so he can call it from subclasses in his units implementation.